rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -884,12 +890,14 @@ static void handle_builtintopic_endpoint(
   bool is_reader)
 {
   uint8_t type_hash[RCUTILS_SHA256_BLOCK_SIZE];
+  memset(type_hash, 0x2c, RCUTILS_SHA256_BLOCK_SIZE);
+
   dds_sample_info_t si;
   void * raw = NULL;
   while (dds_take(reader, &raw, &si, 1, 1) == 1) {
     auto s = static_cast<const dds_builtintopic_endpoint_t *>(raw);
-    RCUTILS_LOG_ERROR("Discovery: %s --- %p",
-      s->topic_name, type_hash);
+    std::string discovered_user_data(s->qos->user_data.value, s->qos->user_data.value + s->qos->user_data.length);
```

Comment:
I guess some special header file brought the definition of `dds_qos_t` into scope here. It is not going to be a problem often, but I'd prefer `dds_qget_user_data` be used, like it is in `parse_user_data`.

In any case, there are two things I need to point out: in principle, the `dds_qos_t` thing is sparse, and so to be safe you need to check whether the QoS actually has been set, by checking the bit in `qos->present` (here, it would be `DDSI_QP_USER_DATA`) — this is handled by `dds_qget_user_data`. Not checking can, in principle, lead to reading uninitialised memory. That said, we do make sure that all the relevant QoS settings have been initialised in the returned object, and so this is not going to be an issue in practice.

The second is that it can be empty, in which case it will be `length` = `0` and `value` = `NULL`. So that, you really should check for.

Reply:
Thanks for the tip, I will use the `dds_qget_user_data` API